# Load RAW Image using Rawpy

In [1]:
import os
import numpy as np
from matplotlib import pyplot as plt
import rawpy
import imageio
from PIL import Image
import array
path = '/home/qingyao/Desktop/Topic 1: Demosaic/AHD Algorithm/7.dng'

# Print out Properties of this RAW image

In [2]:
raw = rawpy.imread(path)
print("size = ",raw.sizes)
print("camera_whitebalance = ",raw.camera_whitebalance)
print("num_colors = ",raw.num_colors)
print("color_desc = ",raw.color_desc)
print("raw_colors = ")
print(raw.raw_colors)
print("raw pattern = ")
print(raw.raw_pattern)
raw_img = raw.raw_image
raw_img_visible = raw.raw_image_visible
print("max pixel value = ",np.max(raw_img_visible))
print("min pixel value = ",np.min(raw_img_visible))
print("data type of pixel = ",type(np.max(raw_img_visible)))

size =  ImageSizes(raw_height=5472, raw_width=7296, height=5472, width=7296, top_margin=0, left_margin=0, iheight=5472, iwidth=7296, pixel_aspect=1.0, flip=3)
camera_whitebalance =  [2.3809523582458496, 1.0, 1.5943223237991333, 0.0]
num_colors =  3
color_desc =  b'RGBG'
raw_colors = 
[[0 1 0 ... 1 0 1]
 [3 2 3 ... 2 3 2]
 [0 1 0 ... 1 0 1]
 ...
 [3 2 3 ... 2 3 2]
 [0 1 0 ... 1 0 1]
 [3 2 3 ... 2 3 2]]
raw pattern = 
[[0 1]
 [3 2]]
max pixel value =  2912
min pixel value =  128
data type of pixel =  <class 'numpy.uint16'>


# Image Properties Analysis
This image has shape 5472 x 7296 without margin. <br>
num of colors is 3 => G1 is equal to G2 <br>
The Bayer pattern is RGBG => filter = 0x94949494 (explained in AHD.c) <br>
Max value 2912 => 12 bit image <br>

# Save into Binary File 
Save the raw data into binary file (data) and check the shape and max pixel value.<br>
It can be seen that the original data is successfully saved into binary file.

In [3]:
a = array.array("H", list(raw_img_visible.flatten()))
with open("./data", "wb") as f:
    a.tofile(f)
b = np.fromfile('./data',dtype=np.uint16)
print("type of binary file = ",type(b))
print("type of data read from binary file = ",type(b[0]))
print("shape of data stored in binary file = ",np.shape(b))
print("shape of original raw data = ",np.shape(raw_img_visible.flatten()))
print("max pixel value in binary file = ",np.max(b))

type of binary file =  <class 'numpy.ndarray'>
type of data read from binary file =  <class 'numpy.uint16'>
shape of data stored in binary file =  (39923712,)
shape of original raw data =  (39923712,)
max pixel value in binary file =  2912
